gap/semigroups/semieunit.gi

```diff
@@ -0,0 +1,715 @@
+  # The elements of yy may not be the first elements of xx.
+  map_y := function(a)
+    return Position(xx, a);
+  end;
+  map_y := MappingByFunction(Domain(yy), Domain(Set([1 .. Size(xx)])), map_y);
+
+  labels := ShallowCopy(yy);
+  Apply(labels, a -> Position(xx, a));
+  y := InducedSubdigraph(x, labels);
+  SetDigraphVertexLabels(y, labels);
+
+  # The semigroup quotient group is not a group object, so find an isomorphism.
+  map_g := IsomorphismPermGroup(grp);
+  map_g := CompositionMapping(
+             SmallerDegreePermutationRepresentation(Image(map_g)), map_g);
+
+  # Create the action of Image(map_g) on x - this is the action of G on X.
+  anti_act := SEMIGROUPS.McAlisterTripleSemigroupConstructAction(xx, map_g);
+  act := function(pt, g)
+    return(anti_act(pt, g ^ -1));
+  end;
+
+  M := McAlisterTripleSemigroup(Image(map_g), act, x, y);
+  iso := function(s)
+    local t;
+    t := s;
+    return MTE(M, (t ^ map_yy) ^ map_y,
+               CongruenceClassOfElement(cong, s) ^ map_g);
+  end;
+  return MappingByFunction(S, M, iso);
+end);
+
+InstallMethod(AsMcAlisterTripleSemigroup,
+[IsSemigroup],
+function(S)
+  return Range(IsomorphismMcAlisterTripleSemigroup(S));
+end);
+
+InstallMethod(IsWholeFamily, "for a McAlister triple semigroup",
+[IsMcAlisterTripleSemigroupElementCollection],
+function(C)
+  return Size(C[1]![3]) = Size(C);
+end);
+
+InstallMethod(ChooseHashFunction, "for McAlister triple semigroup elements",
+[IsMcAlisterTripleSemigroupElement, IsInt],
+function(x, hashlen)
+  local data;
+  data := [ChooseHashFunction(x[1], hashlen),
+           ChooseHashFunction(x[2], hashlen),
+           hashlen];
+  return rec(func := SEMIGROUPS.HashFunctionForMcAlisterTripleSemigroupElements,
+             data := data);
+end);
+SEMIGROUPS.HashFunctionForMcAlisterTripleSemigroupElements := function(x, data)
+  return  (17 * data[1].func(x[1], data[1].data)
+           + data[2].func(x[2], data[2].data)) mod data[3];
+end;
+
+###############################################################################
+# F-inverse Semigroups
+###############################################################################
+# The connected components of the natural partial order will be the
+# congruence classes of the minmum group congruence. Thus we can simply
+# check that precisely one of the sources of the digraph of the natural
+# partial order is in each connected component.
+InstallMethod(IsFInverseMonoid, "for a semigroup",
+[IsSemigroup],
+function(S)
+  local comp, po;
+  if not IsInverseMonoid(S) then
+    return false;
+  fi;
+  po := Digraph(NaturalPartialOrder(S));
+  for comp in DigraphConnectedComponents(po).comps do
+    if not Size(Intersection(comp, DigraphSources(po))) = 1 then
+      return false;
+    fi;
+  od;
+  return true;
+end);
+
+InstallMethod(IsFInverseMonoid, "for a McAlister triple semigroup",
+[IsMcAlisterTripleSemigroup],
+function(S)
+  return IsFInverseSemigroup(S) and IsMonoid(S);
```

Comment:
Swap these conditions around. It's easier to check `IsMonoid` (in fact it's immediate) so it should come first.